ytypes/leafref_test.go

```diff
@@ -0,0 +1,432 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ytypes
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/openconfig/goyang/pkg/yang"
+)
+
+func TestValidateLeafRefData(t *testing.T) {
+	containerWithLeafListSchema := &yang.Entry{
+		Name: "container",
+		Kind: yang.DirectoryEntry,
+		Dir: map[string]*yang.Entry{
+			"leaf-list": {
+				Name:     "leaf-list",
+				Kind:     yang.LeafEntry,
+				Type:     &yang.YangType{Kind: yang.Yint32},
+				ListAttr: &yang.ListAttr{MinElements: &yang.Value{Name: "0"}},
+			},
+			"list": {
+				Name:     "list",
+				Kind:     yang.DirectoryEntry,
+				ListAttr: &yang.ListAttr{MinElements: &yang.Value{Name: "0"}},
+				Key:      "key",
+				Dir: map[string]*yang.Entry{
+					"key": {
+						Name: "key",
+						Kind: yang.LeafEntry,
+						Type: &yang.YangType{Kind: yang.Yint32},
+					},
+					"int32": {
+						Name: "int32",
+						Kind: yang.LeafEntry,
+						Type: &yang.YangType{Kind: yang.Yint32},
+					},
+				},
+			},
+			"int32": {
+				Name: "int32",
+				Kind: yang.LeafEntry,
+				Type: &yang.YangType{Kind: yang.Yint32},
+			},
+			"key": {
+				Name: "key",
+				Kind: yang.LeafEntry,
+				Type: &yang.YangType{Kind: yang.Yint32},
+			},
+			"enum": {
+				Name: "enum",
+				Kind: yang.LeafEntry,
+				Type: &yang.YangType{Kind: yang.Yint64},
+			},
+			"container2": {
+				Name: "container2",
+				Kind: yang.DirectoryEntry,
+				Dir: map[string]*yang.Entry{
+					"int32-ref-to-leaf": {
+						Name: "int32-ref-to-leaf",
+						Kind: yang.LeafEntry,
+						Type: &yang.YangType{
+							Kind: yang.Yleafref,
+							Path: "../../int32",
+						},
+					},
+					"enum-ref-to-leaf": {
+						Name: "enum-ref-to-leaf",
+						Kind: yang.LeafEntry,
+						Type: &yang.YangType{
+							Kind: yang.Yleafref,
+							Path: "../../enum",
+						},
+					},
+					"int32-ref-to-leaf-list": {
+						Name: "int32-ref-to-leaf-list",
+						Kind: yang.LeafEntry,
+						Type: &yang.YangType{
+							Kind: yang.Yleafref,
+							Path: "../../leaf-list",
+						},
+					},
+					"int32-ref-to-list": {
+						Name: "int32-ref-to-list",
+						Kind: yang.LeafEntry,
+						Type: &yang.YangType{
+							Kind: yang.Yleafref,
+							Path: "../../list[key = current()/../../key]/int32",
+						},
+					},
```

Comment:
Consider adding test cases for:

* `leaf-list` containing `leafref`.
* `typedef` that defines a `leafref`.
* (for Y1.1) `union` that defines a `leafref`.

Reply:
Created https://github.com/openconfig/ygot/issues/124